packaging/_parser.py

```diff
@@ -0,0 +1,201 @@
+            tokens.try_read("COLON")
+        if not tokens.try_read("RBRACKET"):
+            tokens.raise_syntax_error(message="Closing square bracket is missing")
+    return extras
+
+
+def parse_specifier(tokens: Tokenizer) -> str:
+    """
+    SPECIFIER: LPAREN (OP + VERSION + COLON)+ RPAREN | OP + VERSION
+    """
+    parsed_specifiers = ""
+    lparen = False
+    if tokens.try_read("LPAREN"):
+        lparen = True
+    while tokens.match("OP"):
+        parsed_specifiers += tokens.read("OP").text
+        if tokens.match("VERSION"):
+            parsed_specifiers += tokens.read("VERSION").text
+        else:
+            tokens.raise_syntax_error(message="Missing version")
+        if tokens.match("COLON"):
+            parsed_specifiers += tokens.read("COLON").text
+    if lparen and not tokens.try_read("RPAREN"):
+        tokens.raise_syntax_error(message="Closing right parenthesis is missing")
+    return parsed_specifiers
+
+
+def parse_quoted_marker(tokens: Tokenizer) -> MarkerList:
+    tokens.try_read("SEMICOLON")
+    return parse_marker_expr(tokens)
+
+
+def parse_marker_expr(tokens: Tokenizer) -> MarkerList:
+    """
+    MARKER_EXPR: MARKER_ATOM (BOOLOP + MARKER_ATOM)+
+    """
+    expression = [parse_marker_atom(tokens)]
+    while tokens.match("BOOLOP"):
+        tok = tokens.read("BOOLOP")
+        expr_right = parse_marker_atom(tokens)
+        expression.extend((tok.text, expr_right))
+    return expression
+
+
+def parse_marker_atom(tokens: Tokenizer) -> MarkerAtom:
+    """
+    MARKER_ATOM: LPAREN MARKER_EXPR RPAREN | MARKER_ITEM
+    """
+    if tokens.try_read("LPAREN"):
+        marker = parse_marker_expr(tokens)
+        if not tokens.try_read("RPAREN"):
+            tokens.raise_syntax_error(message="Closing right parenthesis is missing")
+        return marker
+    else:
+        return parse_marker_item(tokens)
+
+
+def parse_marker_item(
+    tokens: Tokenizer,
+) -> MarkerItem:
+    """
+    MARKER_ITEM: MARKER_VAR MARKER_OP MARKER_VAR
+    """
+    marker_var_left = parse_marker_var(tokens)
+    marker_op = parse_marker_op(tokens)
+    marker_var_right = parse_marker_var(tokens)
+    return (marker_var_left, marker_op, marker_var_right)
+
+
+def parse_marker_var(tokens: Tokenizer) -> MarkerVar:
+    """
+    MARKER_VAR: VARIABLE MARKER_VALUE
+    """
+    if tokens.match("VARIABLE"):
+        return parse_variable(tokens)
+    else:
+        return parse_python_str(tokens)
+
+
+def parse_variable(tokens: Tokenizer) -> Variable:
+
+    env_var = tokens.read("VARIABLE").text.replace(".", "_")
+    if (
+        env_var == "platform_python_implementation"
+        or env_var == "python_implementation"
+    ):
+        return Variable("platform_python_implementation")
+    else:
+        return Variable(env_var)
+
+
+def parse_python_str(tokens: Tokenizer) -> Value:
+
+    if tokens.match("QUOTED_STRING"):
+        python_str = tokens.read().text.strip("'\"")
```

Comment:
```suggestion
        python_str = tokens.read().text[1:-1]
```

Reply:
Would it make sense to use `ast.literal_eval` to handle escape sequences? (I think they are technically allowed...? PEP 345 simply says the value can be “a free string”.)

Reply:
[PEP 508](https://www.python.org/dev/peps/pep-0508/#grammar) specifies a *very* restricted charset for this, actually. But the old parser uses pyparsing's QuotedString, so `ast.literal_eval` would match the old behavior better.
(That is what we want -- where possible, any behavior changes to better match the spec should not be part of *this* change.)